decred/decred/dcr/wire/netaddress.py

```diff
@@ -0,0 +1,222 @@
+
+		Args:
+			service (int): Bitfield which identifies the service(s) to check.
+
+		Returns:
+			bool: True if this peer offers the service(s).
+		"""
+        return self.services & service == service
+
+    def addService(self, service):
+        """
+		Adds service(s) as a supported service.
+
+		Args:
+			service (int): Bitfield which identifies the service(s) to add.
+		"""
+        self.services |= service
+
+
+def newNetAddressIPPort(ip, port, services):
+    """
+	A new NetAddress using the provided IP, port, and supported services with
+	defaults for the remaining fields.
+
+	Args:
+		ip (str or bytes-like): The peer's IP address.
+		port (int): The peer's port.
+		services (int): Bitfield which identifies the services supported by
+			the peer.
+
+	Returns:
+		NetAddress: The peer's NetAddress.
+	"""
+    return newNetAddressTimestamp(int(time.time()), services, ip, port)
+
+
+def newNetAddressTimestamp(stamp, services, ip, port):
+    """
+	A new NetAddress using the provided timestamp, IP, port, and supported
+	services.
+
+	Args:
+		stamp (int): The last time the peer was seen.
+		ip (str or bytes-like): The peer's IP address.
+		port (int): The peer's port.
+		services (int): Bitfield which identifies the services supported by
+			the peer.
+
+	Returns:
+		NetAddress: The peer's NetAddress.
+	"""
+    return NetAddress(ip, port, services, stamp)
+
+
+def newNetAddress(tcpAddr, services):
+    """
+	A new NetAddress using the provided TCPAddr and supported services with
+	defaults for the remaining fields.
+
+	Args:
+		tcpAddr (TCPAddr): The peer's address.
+		services (int): Bitfield which identifies the services supported by
+			the peer.
+
+	Returns:
+		NetAddress: The peer's NetAddress.
+	"""
+    return newNetAddressIPPort(tcpAddr.ip, tcpAddr.port, services)
+
+
+def readNetAddress(b, hasStamp):
+    """
+	Reads an encoded NetAddress from b depending on the protocol version and
+	whether or not the timestamp is included per hasStamp.  Some messages like
+	version do not include the timestamp.
+
+	Args:
+		b (ByteArray): The encoded NetAddress.
+		hasStamp (bool): Whether or not the NetAddress has a timestamp.
+
+	Returns:
+		NetAddress: The decoded NetAddress.
+	"""
+    expLen = 30 if hasStamp else 26
+    if len(b) != expLen:
+        raise DecredError(
+            f"readNetAddress wrong length (hasStamp={hasStamp}) expected {expLen}, got {len(b)}"
+        )
+
+    # NOTE: The Decred protocol uses a uint32 for the timestamp so it will
+    # stop working somewhere around 2106.  Also timestamp wasn't added until
+    # protocol version >= NetAddressTimeVersion
+    stamp = b.pop(4).unLittle().int() if hasStamp else 0
+    services = b.pop(8).unLittle().int()
+    ip = b.pop(16)
+    if ip[:12] == ipv4to16prefix:
+        ip = ip[12:]
+
+    # Sigh.  Decred protocol mixes little and big endian.
+    port = b.pop(2).int()
+
+    return NetAddress(ip=ip, port=port, services=services, stamp=stamp,)
```

Comment:
Not using one of those handy factory functions, how come? :wink: Also, one redundant comma near the end of line.